### PR TITLE
Single element spring damper

### PR DESCRIPTION
**Pull request by morettid**
Meshing the spring-dampers added on a line element breaks into multiple elements, which is incorrect.
This PR adjusts the mesh data by removing all the in-between nodes and elements between the source and target nodes for the spring-damper elements.

### CODE REVIEW

File: stem/model.py [hunk 1]
```diff
@@ -882,6 +873,251 @@ def __initialise_fields(self):
...
+    def __get_maximum_element_id(self) -> int:
+        """
+        Returns the maximum element id within the mesh from the mesh data
+
+        Returns:
+            - int: the maximum element id
+
+        """
+        max_element_id = 0
+        for element_type, mesh_element_info in self.gmsh_io.mesh_data["elements"].items():
```
Comment (brunozc) [inline stem/model.py:939]: element_type not used. can do self.gmsh_io.mesh_data["elements"].values()

File: stem/model.py [hunk 2]
```diff
@@ -882,6 +873,251 @@ def __initialise_fields(self):
...
+        element_ids_search_space = list(model_part.mesh.elements.keys())
+        node_ids_search_space = list(model_part.mesh.nodes.keys())
+
+        # retrieve the connectivity
+        # node -> elements
+        node_to_elements = self.__map_node_to_elements(model_part.mesh)
+
+        # initialise output list
+        line_node_ids = []
+        # initialise a list for end-point we have already encountered n the clustering algorithm
```
Comment (brunozc) [inline stem/model.py:981]: `        # initialise a list for end-point we have already encountered in the clustering algorithm`

File: stem/model.py [hunk 3]
```diff
@@ -882,6 +873,251 @@ def __initialise_fields(self):
...
+            # only consider the end nodes that are not already in the completed_points list
+            if end_node not in completed_points:
+
+                # remove the end node from the list containing the node ids
+                node_ids_search_space.remove(end_node)
+                first_node_id = None
+
+                # if the point is not the end of the cluster, continue until you find the end of the cluster and include
+                # all the line strings
+                while first_node_id not in end_nodes and len(element_ids_search_space) > 0:
```
Comment (brunozc) [inline stem/model.py:994]: alternative to while keeping the same algorithm: 

            for _ in range(len(element_ids_search_space)):
                second_node_id = self.__find_next_node_along_line_elements(
                    first_node_id, element_ids_search_space, node_ids_search_space, node_to_elements,
                    model_part.mesh.elements, node_ids_at_geometry_points
                )

                # Add the end nodes to the list (start node, end node)
                line_node_ids.append([first_node_id, second_node_id])

                # Update the next point for the search
                first_node_id = second_node_id

                # Check if the current point is an end point
                if first_node_id in end_nodes:
                    break

            # Add the end point to completed_points to reduce the search space
            completed_points.add(first_node_id)

File: stem/model.py [hunk 4]
```diff
@@ -882,6 +873,251 @@ def __initialise_fields(self):
...
+                    break
+
+            # reduce search space for next iteration
+            remaining_node_ids.remove(next_node)
+
+            # if the node is one of the nodes of interest, return them, otherwise continue.
+            if next_node in target_node_ids:
+                return next_node
+
+        raise ValueError("Next node along the line cannot be found. Maximum number of iterations exceeded.")
```
Comment (brunozc) [inline stem/model.py:1119]: it is not the maximum number of iteration. it is just that node is not part of the mesh.

File: stem/model.py [hunk 5]
```diff
@@ -882,6 +873,251 @@ def __initialise_fields(self):
...
+        for element in line_elements.values():
+            if element.element_type != "LINE_2N" and element.element_type != "LINE_3N":
+                raise ValueError("Not all elements are line elements.")
+
+        # initialise variables before loop
+        next_node = start_node_id
+
+        # start the search for the connected node
+        max_iterations = len(remaining_element_ids)
+        for i in range(max_iterations):
```
Comment (brunozc) [inline stem/model.py:1091]: `        for _ in range(max_iterations):`

File: stem/model.py [hunk 6]
```diff
@@ -882,6 +873,251 @@ def __initialise_fields(self):
...
+                raise ValueError(f"There is a fork in the mesh at elements: {elements_connected}, the next node along "
+                                 f"the line cannot be found.")
+
+            next_element_id = elements_connected[0]
+
+            # reduce search space for next iteration
+            remaining_element_ids.remove(next_element_id)
+
+            # find the node(s) connected to the element that have not yet been found yet.
+            for node_id in line_elements[next_element_id].node_ids:
```
Comment (brunozc) [inline stem/model.py:1107]: next_node = next(node_id for node_id in line_elements[next_element_id].node_ids
                             if node_id in remaining_node_ids)

**Review by brunozc — CHANGES_REQUESTED** [review]

File: stem/model.py [hunk 7]
```diff
@@ -882,6 +873,251 @@ def __initialise_fields(self):
...
+        new_element_id = self.__get_maximum_element_id() + 1
+
+        # retrieve connectivities and cluster into individual spring-damper elements
+        for mp in self.body_model_parts:
+
+            if (isinstance(mp.material, StructuralMaterial)
+                    and isinstance(mp.material.material_parameters, ElasticSpringDamper)):
+
+                # assert mesh is initialised
+                if mp.mesh is None:
```
[the remark this answers is not in the document]
Comment (aronnoordam) [reply stem/model.py:897]: yes required for mypy. as mp.mesh is also called within this function

File: stem/model.py [hunk 8]
```diff
@@ -882,6 +873,251 @@ def __initialise_fields(self):
...
+                    line_node_ids.append([first_node_id, second_node_id])
+                    # update the next point for the search
+                    first_node_id = second_node_id
+
+                # add the end point to the completed_points in order to reduce the search space
+                completed_points.append(first_node_id)
+        return line_node_ids
+
+    @staticmethod
+    def __map_node_to_elements(mesh: Mesh) -> Dict[int, List[int]]:
```
[the remark this answers is not in the document]
Comment (aronnoordam) [reply stem/model.py:1015]: __map_node_to_elements is also used in  "__get_line_string_end_nodes", thus merging the two functions is inefficient

File: stem/model.py (hunk 3 above)
Comment (aronnoordam) [reply stem/model.py:994]: element_ids_search_space is reducing in size within the loop, i dont think looping over this in a for loop is a smart thing to do

**Review by brunozc — APPROVED** [review]

**Pull request merged**